### ISSUE TITLE
Specify how updates work

### ISSUE THREAD

**Issue by adrianhopebailie** [w3c/manifest#446]
Is there a reason that the spec only defines a declarative way to request that a user agent obtain and process a manifest?

Likewise, it would be useful to have an explicit way to request that the user-agent update its version of the manifest.

**Comment by marcoscaceres** [comment]
@adrianhopebailie yeah, we might need to add an explicit means to request an update to mitigate security issues: that is, we want to be able to show what's changed and give the end-user a chance to review or dismiss the update. 

Excuse the crappy mock-up I slapped together below, but want to illustrate the point:  

![screenshot 2016-04-06 14 16 44](https://cloud.githubusercontent.com/assets/870154/14305698/39373a10-fc02-11e5-9b5e-d06837ce4690.png)

The UA would also need to show more info, like if the display mode or start URL changed, etc. Basically, anything that could be a concern to a user.

**Comment by marcoscaceres** [comment]
@adrianhopebailie, about "installing explicitly", most browser vendors would like to pursue the [installabiltiy signals](https://w3c.github.io/manifest/#installability-signals) route. If that doesn't work out, we might need to add an .requestInstall() API. However, we have resisted that option because we don't want install buttons all over the web.

**Comment by adrianhopebailie** [comment]
What is the logic a user agent should apply to decide if the installation of one manifest is actually a new install or an update/replacement of an existing one? It would seem that this should be related to scope but I'm not sure if that's explicitly defined anywhere?

Put differently, why would you say this?

> we want to be able to show what's changed and give the end-user a chance to review or dismiss the update. 

If I visit the same URL a week apart I get no warnings that the content has changed since the last time I visited?

If the "effect" of a manifest is limited to functions that will not impact security or privacy then the normal controls for managing those risks should suffice surely?

**Comment by marcoscaceres** [comment]
> If I visit the same URL a week apart I get no warnings that the content has changed since the last time I visited?

But you totally do. For instance, Safari shows you the following when you go to bbc.com/news: 

![screenshot_2016-04-08_17_57_08](https://cloud.githubusercontent.com/assets/870154/14377551/96d3cd16-fdb3-11e5-9dff-8242434e31f5.png)

And Chrome: 
![screenshot_2016-04-08_18_07_19](https://cloud.githubusercontent.com/assets/870154/14377785/f99b425c-fdb4-11e5-9c0c-94bf02058e6f.png)

There are a lot of indicators built into browsers to make users aware of what is going on. And, if for instance, you retuned to the bbc.com and their cert had expired, the browser would warn you. 

The same with native apps, like providing either, or both, an accessible history showing what's being updated: 

![screenshot 2016-04-08 18 02 27](https://cloud.githubusercontent.com/assets/870154/14377627/167daf0a-fdb4-11e5-950c-8a37965325e7.png)

The reason it feels different than traditional software updates is that the Web's same origin policy and browser sandboxing protects web pages from meddling with the underlying OS. Web manifest changes that: 

1. it can add an icon to the user's homescreen.  
2. it can set the name of the application. 
3. it can take away the browser's chrome. 
4. it can redirect the browser somewhere unexpected (via start_url)
5. it can hide information about the user in the start URL (or other URLs - if those URLs are rewritten and captured via a service worker, for instance)

> If the "effect" of a manifest is limited to functions that will not impact security or privacy then the normal controls for managing those risks should suffice surely?

But they would impact security and privacy (see all the recently added security and privacy notes in the spec). So doing a "sneaky" switcharoo of the icons and name of the icons that the user is not aware of would be really bad. At least, at Mozilla I know we would be absolutely against just allowing applications to switch any part of the manifest without the user knowing about it (and modifying them).

**Comment by adrianhopebailie** [comment]
I'm sorry but I think of myself as a pretty "experienced" web user and I definitely don't consider, the fact that a browser indicates to me a page is being fetched and loaded, as a hint that that the underlying content has changed. I think that is over-reaching a little.

Are you suggesting that an average user of the web understands the difference between a page loaded from the cache and one fetched from the Web and ALSO understands that the latter MAY mean the code of the Web page is different to the last time they loaded that page?

The fact that `start_url` can change is the only reason all this complex change control is required at all. Wouldn't it be better to bind a manifest to a `start_url`?

Perhaps if you could answer my initial question it would help me understand this better:

> What is the logic a user agent should apply to decide if the installation of one manifest is actually a new install or an update/replacement of an existing one? It would seem that this should be related to scope but I'm not sure if that's explicitly defined anywhere?

It seems to me that it's possible to have a URL that falls under overlapping scopes of two different manifests. In which case, if I land on that URL, which application-context, if any, is loaded?

**Comment by marcoscaceres** [comment]
> It seems to me that it's possible to have a URL that falls under overlapping scopes of two different manifests. In which case, if I land on that URL, which application-context, if any, is loaded?

We haven't worked out what to do about overlapping scopes. Hope to get to that soon.  

However, the problem of requesting an explicit update remains. Relates to https://github.com/w3c/manifest/issues/384

**Comment by boyofgreen** [comment]
Binding to a start URL doesn't quite fit the grander idea of scope either.  Since I may "deep link" into an app, the scope needs to cover all that would be part of the app.  The idea we are trying to solve in when you have an application like gmail, where mail.google.com is the scope, but contacts.google.com should also be in scope.  In contrast, docs.google.com and www.google.com should not be in scope.

**Comment by benfrancis** [comment]
So now that the update text has been removed, as far as the spec is currently concerned a manifest gets set at install time and can never be updated? Although I'm aware of the risk of app identity theft, I'm not sure this is better than what was there before. Seems like a high priority to address updates.

**Comment by marcoscaceres** [comment]
> On 27 Apr 2016, at 9:13 PM, Ben Francis <notifications@github.com> wrote:
> 
> So now that the update text has been removed, as far as the spec is currently concerned a manifest gets set at install time and can never be updated?
> 
Correct.
> Although I'm aware of the risk of app identity theft, I'm not sure this is better than what was there before. Seems like a high priority to address updates.
> 

Yes. I agree.

**Comment by cheshrkat** [comment]
I just ran into this while spiking the use of app manifest. It's great to get onto the device, but the spec definitely needs to talk about the way(s) you can update the content.

**Comment by kenchris** [comment]
AFAIU, the Android team is looking into updating the app data given manifest changes as part of the work on WebAPK. It would be great to have their feedback. @PaulKinlan ?

**Comment by marcoscaceres** [comment]
I think we should limit the declarative features of the manifest as much as possible, particularly ones where we can do better (for developers) by making them programmatic. I believe there is, nevertheless, and as @RobDolinMS suggests, the opportunity for a good mix of collaboration between the browser and the web application: where the browser can automatically check for an update (as service workers do every 24 hours) - or the developer can request an update, through an API.

As I argued in https://github.com/w3c/manifest/issues/446#issuecomment-206112421, the end user needs to be in control of the update process (similarly to how they go through the installation process), because we are asking the user agent to make a pretty scary change at the OS level (the application icon, name, start_url, and possibly other things can change).

**Comment by marcoscaceres** [comment]
@PaulKinlan, feedback would be tremendously useful, particularly if there is going to be some kind of impact on the API surface. It may be we reuse the `onbeforeinstall` infrastructure here to perform the update - which would make most sense.

**Comment by nunoperalta** [comment]
It is really unfortunate we can't update our PWA. This can be considered a stopper in many projects.

manifest.json should have a "version" that can be updated, and if browsers detect that that version number has changed, they can ask the user whether they would like to receive the update or not.

I was asking about this in a StackOverflow question... https://stackoverflow.com/questions/47762022/apply-changes-to-web-app-manifest-on-an-installed-pwa

**Comment by marcoscaceres** [comment]
Version is orthogonal to updating. Nonetheless, update is important and something I hope we solve in 2018.

**Comment by dominickng** [comment]
Note that Chrome now will poll the web manifest of an installed PWA on Android from time to time and automatically update if changes are detected. A more explicit updating mechanism may be better though since Chrome's process is entirely user-transparent.

**Comment by nunoperalta** [comment]
I had tried yesterday to make an update to the app name in the manifest and then change the link to "manifest.json?v=2", which ensures the browser will download the new version. However, the app name did not change.

**Comment by dominickng** [comment]
Changing the manifest URL shouldn't be required (and it may even interfere with the process).

Assuming you have the app installed via the improved add to home screen (i.e. it appears in the Android app list), when you launch the app Chrome will fetch the manifest, parse it, and compare its data to the data we have baked into the app. If things differ, an update is requested and will go through when possible. Hence this being entirely user-transparent, it should just happen in the background.

**Comment by ewsizeit** [comment]
What URL is Chrome using to fetch the manifest? Is it the original url the manifest was fetched from when it was installed or the start_url member?

I see a danger in updating using the manifest presented by the page in start_url. What if you make a mistake in an update, for example a misspelled domain name. All installed apps would update and then have no way of fetching the right manifest again. Basically locking yourself out of your installed app base.

Irregardless of the implementation details, that's great news. We can start producing apps without worrying that they will be impossible to update. The ideal would be to specify this in the standard.

**Comment by frlinw** [comment]
> when you launch the app Chrome will fetch the manifest, parse it, and compare its data to the data we have baked into the app. If things differ, an update is requested
@dominickng I'm afraid it does not work or please give me more details about the update process. I tried with a static name `manifest.json` and with a dynamic one `manifest.[hash].json` and Chrome does not update the manifest, never.
I opened an issue with a screencast example (even if it's not really an issue if there is no spec...) https://bugs.chromium.org/p/chromium/issues/detail?id=806518#c4

I expect a caching strategy similar to others static files. If the manifest href has changed, get & apply the new manifest, if not get the manifest from the cache

**Comment by moneytoo** [comment]
I have a PWA where I have just added "dark mode" in additional to light (and planning on supporting ``prefers-color-scheme`` query). 
It's a single PWA with single manifest but two color modes - light and dark. I would be really nice to be able to update ``background_color`` and ``theme_color`` to avoid light colors of status bar in dark mode or sudden switch from light splashscreen to dark themed app.
I understand that "dark modes" are coming to OS's and browsers and the ``prefers-color-scheme`` is dynamic - it can be set permanently or setup (I would guess) to change based on time of day.

**Comment by tomayac** [comment]
@moneytoo We’re discussing this particular issue in https://github.com/w3c/manifest/issues/758.

**Comment by g-ortuno** [comment]
@moneytoo you can use the `theme-color` meta tag[[1]](https://www.w3.org/TR/appmanifest/#theme_color-member) to override the manifest's `theme_color`. This will only change the status bar though.

**Comment by moneytoo** [comment]
@g-ortuno Thank you for mentioning that. This is what I realized later on and ended up with (along with finding dark/light neutral color for "splash screen").
However I found that I'm unable to override ``theme_color`` from manifest with the color white using ``theme-color`` meta tag - while it works for other colors (that makes me check for standalone mode and use lightgray instead). But that belongs to other issue... (don't know if it's implementation or spec thing)

**Comment by g-ortuno** [comment]
Yeah, that is strange. Might be worth filling a bug on Chrome for that.

**Comment by adrm** [comment]
Hi, my use case also needs some way to dynamically specify a web manifest url. I've seen people using this method https://medium.com/@alshakero/how-to-setup-your-web-app-manifest-dynamically-using-javascript-f7fbee899a61 but reading the spec it should not work, so I prefer not to trust it. I was expecting to be able to do what I am already doing with favicons, which is inserting a `<link>` element dinamically, and every browser seems to update the favicon as expected. Is there some way to do this right now?

**Comment by aarongustafson** [comment]
> Hi, my use case also needs some way to dynamically specify a web manifest url. I've seen people using this method https://medium.com/@alshakero/how-to-setup-your-web-app-manifest-dynamically-using-javascript-f7fbee899a61 but reading the spec it should not work, so I prefer not to trust it. I was expecting to be able to do what I am already doing with favicons, which is inserting a `<link>` element dinamically, and every browser seems to update the favicon as expected. Is there some way to do this right now?

@adrm Do you need to do it post-render? Could you do it server-side? What’s the use case? More details would help with a recommendation.

**Comment by adrm** [comment]
@aarongustafson we are shipping an app as static files through Netlify, and we serve it from multiple domains, where it customizes itself with different names, logos and theme colors depending on the host. The app gets the host from the client and tells the server where it is running. The server do not have other way to have this information because calls to the server are proxied through Netlify so all calls seems to come from them. 

This works wonders in every aspect since all network calls are controlled by javascript and we can easily inject a custom header with the host. The only requests we can’t control this way are the favicon and manifest since the app does not know the host where it’s running until runtime. We checked and the favicon could be loaded dynamically inserting the link tag in head with a parameter signaling the host to the server, but it seems we cannot do the same for the manifest.

I think it would be enough and expected by users if it worked as the favicon tag and we could insert it on runtime. I suppose there is no widely supported workaround right now, but if there are any suggestions I would love to hear them. Thanks.

**Comment by christianliebel** [comment]
Related Version History proposal by the MS Edge team: 
https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/master/VersionHistory/explainer.md

.cc @aarongustafson

**Comment by nhoizey** [comment]
And a more specific need for Manifest and Service Worker/app update sync:
https://github.com/MicrosoftEdge/MSEdgeExplainers/issues/216

**Comment by marcoscaceres** [comment]
We don't have a conclusion for this, but we should figure this out as part of CR... at least check what Safari and Chrome doe here. We don't do anything AFAIK in Gecko.

**Comment by marcoscaceres** [comment]
@kenchris will investigate what happens in Chrome...

**Comment by mgiuca** [comment]
@alancutter has been working on the Chrome manifest update logic, and should be able to answer questions on this.

**Comment by alancutter** [comment]
Chrome compares the latest manifest against what is currently installed throttled to once per day. If changes are found it will wait until all app windows are closed then update its web app installation to reflect those changes (app name excluded).

I can imagine sites that want control over this behaviour would like a checkForManifestUpdates() function and a beforemanifestupdate event that can be event.preventDefault()ed.

**Comment by g-ortuno** [comment]
I suspect we do something differently on Android though.

**Comment by alancutter** [comment]
Android does the same except the check is done on a central server and manifest updates are pushed out globally. This would require major changes to support any kind of update API.

**Comment by dominickng** [comment]
@alancutter @g-ortuno the manifest fetching and parsing for updates is still done on-device on Android, with parsed values and icons being sent to the server to compute whether or not an update is needed based on those. Supporting explicit updating would probably just require some extra flag to "force an update", which may already exist.

The implementation side should be orthogona though - here we should be answering the question of whether this is a sensible and desirable functionality that should be supported.

**Comment by marcoscaceres** [comment]
Some documentation for Chrome for Android: https://web.dev/manifest-updates/

**Comment by evanstade** [comment]
> Version is orthogonal to updating.

Can you explain what you mean by this @marcoscaceres ? It seems straightforward enough to define a rule such as, "if the user agent finds that the version in the manifest is higher than the version at last install/update time, it MAY apply changes found in the manifest to the installed app"

I imagine a version field would _also_ be useful for orthogonal things, like helping developers identify the code that's being run (e.g. on a user's system when a bug is reported).

> Chrome compares the latest manifest against what is currently installed throttled to once per day. If changes are found it will wait until all app windows are closed then update its web app installation to reflect those changes

(Desktop) Chrome is forced to not just compare the manifest to the last seen manifest, but also re-download and compare all images every time, which is a fairly intensive operation. Hence the throttling. But if Chrome were able to skip the update by avoiding this step and only updating after an explicit update signal (via version comparison), the throttling might not be necessary. The throttling is a known/common source of confusion, to the extent that we have documentation that we frequently point confused developers to.

**Comment by dmurph** [comment]
Quick thoughts on version:
* Would be a nice hint
* Relying on it would obviously break existing use cases, but we could throttle non-hint cases to reduce resources, and guarantee resources
* Possibly a privacy risk - version could be a user id. Would have to be not accessible, and even then, it reveals some info about the value when the manifest updates?
  * but I guess we could reset version on browsing data cleared?


I see value as an improvement of update predictability.

**Comment by aarongustafson** [comment]
Related: https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/VersionHistory/explainer.md

**Comment by marcoscaceres** [comment]
@evanstade wrote: 
> Can you explain what you mean by this @marcoscaceres ? It seems straightforward enough to define a rule such as, "if the user agent finds that the version in the manifest is higher than the version at last install/update time, it MAY apply changes found in the manifest to the installed app"

Sure, the problem is: how do you compute "higher"? Do you do a simple ascii comparison? Or would you require implementing something like sem-ver, which would lead to a lot of complexity (e.g., "01.01.10-beta-1" GT "1.1.10-beta-1""). The best we can do is "is it different"?. What if people don't want to use sem-ver? and so on... 

> I imagine a version field would also be useful for orthogonal things, like helping developers identify the code that's being run (e.g. on a user's system when a bug is reported).

Sure, but that's equally handled by "version" being an opaque string (as in, "oh, looks like that's good ol' version 'vva.234fad' at it again").

**Comment by aspiers** [comment]
I'm strongly in favour of a versioning scheme to enable predictable updates.  As we all know, versioning is a solution to this problem which is used and well-understood across the whole industry, and it should be easy to retain backwards compatibility.

I would suggest using semver unless there are compelling alternatives or reasons not to (which seem unlikely to me at this point).  Semver is a very well-trodden path and avoids reinventing wheels.  I don't think it's compelling to say that semver leads to a lot of complexity, because it's essentially a very simple concept which is already well understood by many developers and already has multiple implementations.

In other words, update mechanisms via versioning is already a solved problem, so I think it makes sense to treat it as one rather than looking for a new solution.  Reinventing wheels would mean a whole new updates and/or versioning scheme for developers to learn, even though it would only exist in this one place.
